nilearn/_utils/niimg_conversions.py

```diff
@@ -62,18 +135,40 @@ def check_niimg(niimg, atleast_4d=False):
 
     Its application is idempotent.
     """
-    # If it's an iterator, it's a 4d image
+    # in case of an iterable
     if hasattr(niimg, "__iter__") and not isinstance(niimg, _basestring):
-        return check_niimg_4d(niimg)
+        if ensure_ndim == 3:
+            raise TypeError(
+                "Data must be a 3D Niimg-like object but you provided a %s."
+                " See http://nilearn.github.io/building_blocks/"
+                "manipulating_mr_images.html#niimg." % type(niimg))
+        if return_iterator:
+            return _iter_check_niimg(niimg, ensure_ndim=ensure_ndim)
+        return concat_niimgs(niimg, ensure_ndim=ensure_ndim)
 
     # Otherwise, it should be a filename or a SpatialImage, we load it
     niimg = load_niimg(niimg)
 
+    if ensure_ndim == 3 and len(niimg.shape) == 4 and niimg.shape[3] == 1:
```

Comment:
I still think a Is4D() function might be useful. We do this test a number of times across the code.